### DISCUSSION TITLE
What is the intended way to use DB with aiogram?

### DISCUSSION THREAD

**Discussion by kuk** [aiogram/aiogram#967] [General]
I query DynamoDB from aiogram bot. What is the intended way to use DB from aiogram?

## `dispatcher.storage`

As I understand `dispatcher.storage` is only to store user state for FSM + limit requests rate. I need to do generic queries like "find newest post in posts table"

## Implement singleton DB similar to Bot and Dispatcher

I do not like global DB object. Seems hard to test

```python
class DB(ContextInstanceMixin):
  ...


@db.message_handler
async def start_command(message):
  db = DB.get_current()
  result = await db.query(...)


async def on_startup(_):
  db = DB(...)
  await db.connect(...)


async def on_shutdown(_):
  db = DB.get_current()
  await db.close()
```

## My current solution with context

Extra code, seem non native to aiogram

```python
async def handle_start_command(context, message):
  result = await context.db.query(...)


def setup_handlers(context):
    context.dispatcher.register_message_handler(
        context.handle_start_command,
    )


async def on_startup(context, _):
    await context.db.connect()


async def on_shutdown(context, _):
    await context.db.close()


class BotContext:
    def __init__(self):
        self.bot = Bot(token=BOT_TOKEN)
        self.dispatcher = Dispatcher(self.bot)
        self.db = DB()


BotContext.handle_start_command = handle_start_command
BotContext.setup_handlers = setup_handlers

BotContext.on_startup = on_startup
BotContext.on_shutdown = on_shutdown
```

**Comment by salko-ua** [comment]
you can use any database with aiogram, the easiest way I think would be to use the aiosqilte library
https://aiosqlite.omnilib.dev/en/stable/

**Comment by JrooTJunior** [comment]
Both ways from your examples is wrong.

I'll preffer to not say how to solve your issue on legacy 2.x version, it's time to migrate to the latest stable version 3.x and solve this issue on it with described way in the documentation.

https://docs.aiogram.dev/en/dev-3.x/dispatcher/dependency_injection.html
https://docs.aiogram.dev/en/dev-3.x/dispatcher/middlewares.html#function-based

Solution for 2.x should be similar.
